rpc/dialer.go

```diff
@@ -241,6 +243,10 @@ func dialDirectGRPC(ctx context.Context, address string, dOpts dialOptions, logg
 	dialOpts := []grpc.DialOption{
 		grpc.WithBlock(),
 		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(MaxMessageSize)),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                15 * time.Second, // a little extra buffer to try to avoid ENHANCE_YOUR_CALM
+			PermitWithoutStream: true,
```

Comment:
I think this means the device will send a packet every 15 seconds

These are HTTP2 pings per [grpc-go readme](https://github.com/grpc/grpc-go/blob/bb49a8868a2a48de9321a5ad6f0ab4ccfd805f77/Documentation/keepalive.md?plain=1#L3), so this shouldn't create extra log noise

the ClientParameters struct has Time and Timeout, so I think your config means the actual timeout is 35 seconds (15 + 20)?

also, I'm assuming RDK is used to having a transport that never goes away. now that this can close, does RDK reconnect cleanly with backoff? (i.e. not a try-fail flood).

Reply:
35 seconds is correct and yeah it reconnects as you'd expect it to. There used to be keep alive but we removed it

Reply:
The RDK won't/can't see a disconnect in this way; it never has. The underlying load balancer does the connection logic

Reply:
> underlying load balancer

ah you're talking about grpc client-side load balancing like in [grpc docs here](https://grpc.io/blog/grpc-load-balancing/#proxy-or-client-side), my brain kept hearing 'load balancer' and thinking server

reconnect logic lives in the grpc library, makes sense https://github.com/grpc/grpc-go/blob/bb49a8868a2a48de9321a5ad6f0ab4ccfd805f77/clientconn.go#L584-L587

Reply:
commit removing old one https://github.com/viamrobotics/goutils/commit/ce35a14fc60f8a303d7479e61c6051dca2854688

Reply:
Yeah I don't buy that Google GCP load balancers for grpc are sending any keep alive params that are helpful for us. Otherwise we'd see faster connection reconnects

Reply:
So with 35 seconds it feels harmless. Happy to adjust it though

Reply:
oh yeah no opinion on what the default should be, just wrapping my head around this part of the codebase

Reply:
> sending any keep alive params that are helpful for us

^ what do you mean by this? does the cloud side of these connections send down config when a connection is established which the grpc client then uses to configure tcp keepalive?

Reply:
> > sending any keep alive params that are helpful for us
> 
> ^ what do you mean by this? does the cloud side of these connections normally send down config when a connection is established which the grpc client then uses to configure tcp keepalive?

I misspoke, my bad. Both sides are just supposed to agree on the values, but GCP doesn't publish what those are. I believe I removed the keep alives because we were seeing a lot of GOAWAYS, which the grpc library handles by backing off future keep alives.

Reply:
also can you dump repro instructions somewhere for 'too many attempts' error when switching interfaces? I'll diff how this behaves with LB vs directly to the backend service, from there can experiment with LB config changes

Reply:
Yeah, just run a blank config from app.viam.com and switch between net ifaces being up and down that have internet access. In my test, I used nmcli to switch wifi networks.